### PR TITLE
Parse Plane Issues

### PR DESCRIPTION
**Pull request by AlwinEgger**

### CODE REVIEW

**Comment by genox** [comment]
@AlwinEgger The yarn lock file changed, did you rebase on the latest commit on main?

File: backend/app/actors/scheduler/plane/PlaneApiService.scala [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      auth: ServiceAuthentication,
+      executionContext: ExecutionContext): Future[Seq[PlaneLabel]]
+}
+
+class PlaneApiServiceImpl(override val ws: WSClient,
+                          override val config: ServiceConfiguration)
+    extends PlaneApiService
+    with ApiServiceBase {
+
+  val findIssuesUrl = s"/api/v1/workspaces/tegonal-intern/projects/%s/issues/?"
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneApiService.scala]: we should extract the workspace to the config

File: backend/app/actors/scheduler/plane/PlaneApiService.scala [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                 maxResults: Option[Int] = None)(implicit
+      auth: ServiceAuthentication,
+      executionContext: ExecutionContext): Future[PlaneIssuesSearchResult] = {
+
+    val currentPage       = page.getOrElse(0)
+    val currentMaxResults = maxResults.getOrElse(100)
+
+    val params = getParamList(
+      Some(paramString),
+      getParam("cursor", s"${currentMaxResults}:${currentPage}:0"),
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneApiService.scala:97]: not sure how plane react if we send here a value > 100. Maybe we should limit it to 100 as that's the max returned results by the server

File: backend/app/actors/scheduler/plane/PlaneApiService.scala [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      getParam("per_page", currentMaxResults))
+
+    val url = findIssuesUrl.format(projectId) + params
+    logger.debug(s"findIssues: $url")
+    getSingleValue[PlaneIssueWrapper](url).map { pair =>
+      PlaneIssuesSearchResult(
+        pair._1.results,
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneApiService.scala]: I would used named arguments here for a better understanding.

File: backend/app/actors/scheduler/plane/PlaneModels.scala [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    count: Int,
+    total_pages: Int,
+    total_results: Int,
+    extra_stats: Option[String],
+    results: Seq[PlaneIssue],
+)
+
+case class PlaneLabel(
+    id: String,
+    // created_at: DateTime, // Nannosecond format not parsable by joda time "2024-08-27T14:33:01.364694+02:00"
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneModels.scala]:
```suggestion
    // created_at: DateTime, // Nanosecond format not parsable by joda time "2024-08-27T14:33:01.364694+02:00"
```

File: backend/app/actors/scheduler/plane/PlaneModels.scala [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    color: String,
+    group: String
+)
+
+case class PlaneProject(
+    id: String,
+    identifier: String,
+    name: String,
+    cover_image: String,
+// icon_prop: IconProp,
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneModels.scala]: nitpick: stange idention

File: backend/app/actors/scheduler/plane/PlaneModels.scala [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+object PlaneIssueWrapper {
+  implicit val jsonFormat: Format[PlaneIssueWrapper] =
+    Json.format[PlaneIssueWrapper]
+}
+
+object PlaneLabel {
+  implicit val jsonFormat: Format[PlaneLabel] = Json.format[PlaneLabel]
+}
+
+object PlaneState {
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneModels.scala:120]: seems to be unused

File: backend/app/actors/scheduler/plane/PlaneModels.scala [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+  implicit val jsonFormat: Format[PlaneState] = Json.format[PlaneState]
+}
+
+object PlaneProject {
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneModels.scala:124]: seems to be unused

File: backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        log.debug(s"andThen:restart time $s")
+        cancellable = Some(
+          context.system.scheduler
+            .scheduleOnce(settings.checkFrequency milliseconds, self, Parse))
+      }
+  }
+
+  private def toPlaneIssueTag(issue: PlaneIssue): PlaneIssueTag = {
+
+    val nameTag = projectSettings.tagConfiguration.useTitle match {
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala]: `Trivial match can be simplified ` to 
```
 val nameTag = if (projectSettings.tagConfiguration.useTitle) {
      Some(SimpleTag(TagId(issue.name)))
    } else {
      None
    }
```

File: backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+      case false => None
+      case _     => Some(SimpleTag(TagId(issue.name)))
+    }
+
+    val labelTags = projectSettings.tagConfiguration.useLabels match {
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala]: same here

File: backend/app/core/db/InitialDemoDataLoader.scala [hunk 10]
```diff
@@ -165,6 +170,60 @@ class InitialDemoDataLoader @Inject() (
...
+                             maxResults = Some(100),
+                             None,
+                             Some(p.key),
+                             PlaneTagConfiguration(useLabels = true, labels))
+      )
+    })
+
+    val planeConfig = PlaneConfig(
+      PlaneConfigId(),
+      "organize.tegonal.com - Internal Stuff",
```
Comment (toggm) [inline backend/app/core/db/InitialDemoDataLoader.scala]: I would either remove plane integration from demo data, make it configurable for other use cases as well (name) or create another tegonal specific data loader which we reference in our test instance.

File: backend/app/models/PlaneConfig.scala [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    extends BaseBSONObjectId
+
+case class PlaneSettings(checkFrequency: Long)
+
+case class PlaneTagConfiguration(useLabels: Boolean,
+                                 labelFilter: Set[String],
+                                 useMilestone: Boolean = false,
+                                 useTitle: Boolean = false)
+
+case class PlaneProjectSettings(planeProjectId: String,
```
Comment (toggm) [inline backend/app/models/PlaneConfig.scala]: here we would need the workspace id as well

File: backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+              .map(l => SimpleTag(TagId(l.name)))
+          case None => Seq()
+        }
+    }
+
+    val tags =
+      nameTag.map(t => labelTags :+ t).getOrElse(labelTags)
+
+    val issueLink =
+      s"https://organise.tegonal.com/tegonal-intern/projects/${issue.project.id}/issues/${issue.id}"
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala]: here we need a more dynamic configuration including host and workspace or maybe add somethink like external_base_url

File: backend/app/actors/scheduler/plane/PlaneApiService.scala [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+      executionContext: ExecutionContext): Future[Seq[PlaneLabel]]
+}
+
+class PlaneApiServiceImpl(override val ws: WSClient,
+                          override val config: ServiceConfiguration)
+    extends PlaneApiService
+    with ApiServiceBase {
+
+  val findIssuesUrl = s"/api/v1/workspaces/tegonal-intern/projects/%s/issues/?"
+  val fetchLabelUrl = s"/api/v1/workspaces/tegonal-intern/projects/%s/labels/?"
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneApiService.scala]: nitpick: could be a private  member

**Review by toggm — CHANGES_REQUESTED** [review]

File: backend/app/actors/scheduler/plane/PlaneApiService.scala (hunk 13 above)
Comment (AlwinEgger) [reply backend/app/actors/scheduler/plane/PlaneApiService.scala]: Done

File: backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala (hunk 12 above)
Comment (AlwinEgger) [reply backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala]: Done

File: backend/app/models/PlaneConfig.scala (hunk 11 above)
Comment (AlwinEgger) [reply backend/app/models/PlaneConfig.scala]: Done

File: backend/app/core/db/InitialDemoDataLoader.scala (hunk 10 above)
Comment (AlwinEgger) [reply backend/app/core/db/InitialDemoDataLoader.scala]: I removed it from demo data

File: backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala (hunk 9 above)
Comment (AlwinEgger) [reply backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala]: Done

File: backend/app/actors/scheduler/plane/PlaneModels.scala (hunk 4 above)
Comment (AlwinEgger) [reply backend/app/actors/scheduler/plane/PlaneModels.scala]: Done

File: backend/app/actors/scheduler/plane/PlaneApiService.scala (hunk 3 above)
Comment (AlwinEgger) [reply backend/app/actors/scheduler/plane/PlaneApiService.scala]: Done

File: backend/app/actors/scheduler/plane/PlaneApiService.scala (hunk 2 above)
Comment (AlwinEgger) [reply backend/app/actors/scheduler/plane/PlaneApiService.scala:97]: added 100 as the max for this integer

File: backend/app/actors/scheduler/plane/PlaneApiService.scala (hunk 1 above)
Comment (AlwinEgger) [reply backend/app/actors/scheduler/plane/PlaneApiService.scala]: Done

File: backend/conf/demo.conf [hunk 14]
```diff
@@ -23,4 +23,9 @@ akka.contrib.persistence.mongodb.mongo.mongouri = "mongodb://localhost:27017/las
 akka.contrib.persistence.mongodb.mongo.mongouri = ${?MONGODB_URI}
 
 lasius.title = Lasius
-lasius.instance = Demo
\ No newline at end of file
+lasius.instance = Demo
+
+lasius.plane.api_key = "plane_api_08ba10840ad34801baab20fba68a784c"
```
Comment (toggm) [inline backend/conf/demo.conf]: oh, oh...that's a pre-christmas gift ;-)

File: backend/app/actors/scheduler/plane/PlaneApiService.scala (hunk 2 above)
Comment (toggm) [reply backend/app/actors/scheduler/plane/PlaneApiService.scala:97]: `max` returns the maximum number of both integers which means we are now not able to lower the maxResults to less than 100. Guess you should use `min` here instead.

File: backend/app/actors/scheduler/plane/PlaneApiService.scala (hunk 3 above)
Comment (toggm) [reply backend/app/actors/scheduler/plane/PlaneApiService.scala]: I still don't see named arguments.

File: backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala [hunk 15]
```diff
@@ -81,7 +81,30 @@ class PlaneTagParseWorker(wsClient: WSClient,
   }
 
   val parsing: Receive = { case Parse =>
-    loadIssues(0, None)
+    val labelF = loadLabels()
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala]: nitpick: what does the `F` stand for?

File: backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala [hunk 16]
```diff
@@ -81,7 +81,30 @@ class PlaneTagParseWorker(wsClient: WSClient,
...
+    // load labelIds from projectSettings.tagConfiguration.includeOnlyIssuesWithLabels
+    val labelIds = labelF
+      .map { labels =>
+        val excludeLabels = projectSettings.tagConfiguration.includeOnlyIssuesWithLabels
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneTagParseWorker.scala]: looks like a naming problem: `exclude` vs  `ìncludeOnly`, right?

**Review by toggm — CHANGES_REQUESTED** [review]

File: backend/app/actors/scheduler/plane/PlaneModels.scala (hunk 7 above)
Comment (toggm) [reply backend/app/actors/scheduler/plane/PlaneModels.scala:124]: still unused

File: backend/app/actors/scheduler/plane/PlaneApiService.scala (hunk 2 above)
Comment (AlwinEgger) [reply backend/app/actors/scheduler/plane/PlaneApiService.scala:97]: good catch.

File: backend/app/actors/scheduler/plane/PlaneApiService.scala (hunk 3 above)
Comment (AlwinEgger) [reply backend/app/actors/scheduler/plane/PlaneApiService.scala]: your reference is outdated.

File: backend/conf/demo.conf (hunk 14 above)
Comment (AlwinEgger) [reply backend/conf/demo.conf]: noooo!

File: backend/app/actors/scheduler/plane/PlaneApiService.scala [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+      getParam("cursor", s"${currentMaxResults}:${currentPage}:0"),
+      getParam("per_page", currentMaxResults),
+      if(includeOnlyIssuesWithLabelsIds.isEmpty) None else getParam("labels", includeOnlyIssuesWithStateIds.mkString(",")),
+      if(includeOnlyIssuesWithStateIds.isEmpty) None else getParam("state", includeOnlyIssuesWithStateIds.mkString(",")))
+
+    val url = findIssuesUrl.format(workspace, projectId) + params
+    logger.debug(s"findIssues: $url")
+    getSingleValue[PlaneIssueWrapper](url).map { case (planeIssueWrapper, _) =>
+      PlaneIssuesSearchResult(
+        planeIssueWrapper.results,
```
Comment (toggm) [inline backend/app/actors/scheduler/plane/PlaneApiService.scala]: Here named arguments would make sense.

**Review by toggm — APPROVED** [review]

**Pull request merged**